app_dart/lib/src/datastore/datastore_wrapper.dart

```diff
@@ -0,0 +1,77 @@
+// Copyright 2020 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+import 'dart:math';
+import 'package:gcloud/db.dart';
+import 'package:cocoon_service/src/foundation/utils.dart';
+import '../datastore/cocoon_config.dart';
+
+/// Wrapper class for datastore operations.
+///
+/// Datastore operations generate high rates of errors related to
+/// Transaction and/or GRPC problems. Fixing this required to add retries
+/// to all the database operations causing 4 or 5 levels of nesting. This
+/// class was created to simplify all the database related operations to a
+/// single nesting level in most cases or two at most while still providing
+/// retry logic.
+class DatastoreWrapper {
```

Comment:
I thought `datastore.dart` is already the wrapper class for Datastore code. What's the reasoning for having these in a separate file?

Reply:
datastore.dart is the dart implementation for AppEngine datastore. This wrapper class is to provide a consistent retry on failure behavior to the cocoon backend.

This class will also be used to simplify the transactional code everywhere in the cocoon backend removing the several nesting level required if you are using datastore.dart directly.

This is part 1 of many other PRs that will simplify the code in the handlers.

Reply:
`import 'package:gcloud/db.dart';` is the Dart implementation of Datastore. `datastore.dart` is Cocoon's wrapper for that library. Granted, there is datastore code sprinkled throughout the codebase which should be moved to `datastore.dart`.

It seems more idiomatic to call `datastore.insert(...)` where `datastore` already exists in the codebase versus adding a new variable and calling it like `datastoreWrapper.insert(...)`.

Reply:
@CaseyHillers Not sure which datastore are you referring to, the one that is everywhere in the backend code comes from import 'package:gcloud/datastore.dart'; and all the handlers are using Config.db rather than datastore directly.

@keyonghan why we'll need to change the queries, sharding the inserts was already happening in most places where more than 1 model was being inserted, would you mind expanding on that?

Reply:
Oh, what I meant is changing the reference point from datastore to the wrapper for all APIs.

Reply:
yes, I made all those changes locally but will send in separate reviews to simplify the reviews and minimize the impact during the releases.

Reply:
FYI `package:gcloud/db.dart` is just another layer on top of `package:gcloud/datastore.dart`, so they're just the same code :)
https://github.com/dart-lang/gcloud/blob/master/lib/db.dart

Reply:
To clarify my earlier point about this change being more idiomatic in the DatastoreService, see https://github.com/flutter/cocoon/blob/master/app_dart/lib/src/request_handlers/update_benchmark_targets.dart

```
    final DatastoreService datastore = datastoreProvider();
    /// ...
    await datastore.db.commit(inserts: <TimeSeries>[timeSeries]); // could be changed to just datastore.commit(...);
```

Reply:
To clarify what you are suggesting is to move the content of datastore_wrapper to service/datastore.dart and remove the new variable and db from Config. Is that correct?

Reply:
Yes

Reply:
Code moved to service/datastore.